### DISCUSSION TITLE
Why can't it find the reference for the items used in macro?

### DISCUSSION THREAD

**Discussion by cpcopy** [clangd/clangd#2081] [Q&A]
These are the sample codes.
test.h
```
#include <stdio.h>

typedef struct{
    int aa;
    int bb;
    int cc;
} TEST_T;

#define VAR_A(V)  V.aa
```

test.c
```
#include "test.h"

TEST_T var = {0};

int main()
{
    VAR_A(var) = 1;
}
```
If I move the cursor to `aa `in `#define VAR_A(V)  V.aa` and try to find the reference, it will be failed. But it can correctly go to the definition in the struct. And it can find the reference for `aa` if cursor is on `aa` in struct. My doubt is why it is not able to find the reference for `aa` in `#define VAR_A(V)  V.aa` since it can correctly find the definition. One possibility I can think out is there may be several same `aa` in different structure and it cannot figure out which definition is right one used in macro. But, can it work if the variable is the unique one?

**Comment by HighCommander4** [comment]
The reason that find-references does not work is explained in https://github.com/clangd/clangd/issues/346.

The reason go-to-definition works is that we have a textual fallback for go-to-definition, implemented in https://github.com/clangd/clangd/issues/241.

A way to illustrate that the textual fallback is being used is to introduce some other symbol with the same name:

```c++
typedef struct{
    int aa;
    int bb;
    int cc;
} TEST_T;

int aa;

#define VAR_A(V)  V.aa
```

Now, go-to-definition offers **two** targets, the global variable `aa` and `TEST_T::aa`, indicating that clangd has no idea which declaration is being referenced by the symbol and it's just returning any declaration with the name `aa`.

**Reply by cpcopy** [reply]
Thanks for the detailed explanation. It is same to what I suspect. 
Tried with Visual Studio, it lists the groups for different definitions. Can clangd do similar thing? If it only has one definition target, it will work same as common find-reference.
![image](https://github.com/clangd/clangd/assets/80554779/b8f0f93a-1ae0-47e3-951d-f16d2e60a9de)

**Reply by HighCommander4** [reply]
> Tried with Visual Studio, it lists the groups for different definitions. Can clangd do similar thing?

There does not seem to be any support in the Language Server Protocol's [find references operation](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_references) for grouping the results in any way.

So, I think such a feature would need to start by requesting a protocol change at https://github.com/microsoft/language-server-protocol/issues/, and then clangd could implement the revised protocol (and clients like vscode would also need to implement the revised protocol to visualize the groups in their UI).

**Reply by cpcopy** [reply]
I see. Thanks. Tried with vscode, the behavior is same with clangd.
